File: backend/recipe_objects/UserRecipeManager.js
// recipe_objects/UserRecipeManager.js

const fs = require('fs');
const path = require('path');
const Recipe = require('./Recipe');
const Ingredient = require('./Ingredient');

class UserRecipeManager {
  constructor() {
    // Define the data file location; adjust the path if needed.
    this.dataFile = path.join(__dirname, '..', 'data', 'recipes.json');
    this.recipeList = [];
    this.ingredientList = [];
    this.loadData();
  }

  /**
   * Adds a Recipe object to the recipeList if it doesn't already exist by name,
   * and also adds its ingredients.
   * @param {Recipe} recipe
   */
  addRecipeToList(recipe) {
    if (!(recipe instanceof Recipe)) return;

    // Check if a recipe with the same name already exists.
    if (!this.recipeList.some(existingRecipe => existingRecipe.name === recipe.name)) {
      this.recipeList.push(recipe);

      // Loop through each ingredient in the recipe and add it.
      for (const ingredientName in recipe.ingredientDictionary) {
        if (Object.prototype.hasOwnProperty.call(recipe.ingredientDictionary, ingredientName)) {
          const ingData = recipe.ingredientDictionary[ingredientName];
          // Create an Ingredient instance.
          const ingredient = new Ingredient(ingredientName, ingData.price, ingData.unit);
          this.addIngredientToList(ingredient);
        }
      }

      // Save the updated data.
      this.saveData();
    } else {
      console.log(`Recipe '${recipe.name}' already exists in the list.`);
    }
  }

  /**
   * Adds an Ingredient object to the ingredientList if it doesn't already exist by name.
   * @param {Ingredient} ingredient
   */
  addIngredientToList(ingredient) {
    if (!(ingredient instanceof Ingredient)) return;

    if (!this.ingredientList.some(existingIngredient => existingIngredient.name === ingredient.name)) {
      this.ingredientList.push(ingredient);
      this.saveData();
    } else {
      console.log(`Ingredient '${ingredient.name}' already exists in the list.`);
    }
  }

  /**
   * Saves recipes and ingredients to a JSON file.
   */
  saveData() {
    // Convert recipes and ingredients to JSON-friendly objects using toDict() methods.
    const data = {
      recipes: this.recipeList.map(recipe => recipe.toDict()),
      ingredients: this.ingredientList.map(ingredient => ingredient.toDict())
    };

    // Ensure the directory exists (it creates the directory if it doesn't exist).
    fs.mkdirSync(path.dirname(this.dataFile), { recursive: true });

    // Write the JSON data to the file.
    fs.writeFileSync(this.dataFile, JSON.stringify(data, null, 4), 'utf8');
    console.log('Data saved successfully.');
  }

  /**
   * Loads recipes and ingredients from a JSON file, if it exists.
   */
  loadData() {
    if (fs.existsSync(this.dataFile)) {
      try {
        const fileData = fs.readFileSync(this.dataFile, 'utf8');
        const data = JSON.parse(fileData);

        // Convert plain objects back into Recipe and Ingredient instances.
        this.recipeList = (data.recipes || []).map(recipeData => Recipe.fromDict(recipeData));
        this.ingredientList = (data.ingredients || []).map(ingData => Ingredient.fromDict(ingData));
        console.log('Data loaded successfully.');
      } catch (error) {
        console.error('Error loading data:', error);
      }
    }
  }

  getAllRecipes() {
    return this.recipeList;
  }

  // Other methods for adding recipes/ingredients would go here...
}

module.exports = UserRecipeManager;

File: backend/UserRecipeManager.py
import json
import os

from backend.recipe_objects.Recipe import Recipe
from backend.recipe_objects.Ingredient import Ingredient


class UserRecipeManager:
  DATA_FILE = "data/recipes.json"  # Define file path

  def __init__(self):
    self.recipe_list = []
    self.ingredient_list = []
    self.load_data()  # Load saved data on initialization

  def add_recipe_to_list(self, recipe):
    if not isinstance(recipe, Recipe):
      return

    """Adds a recipe object to the list if it doesn't already exist by name."""
    if not any(existing_recipe.name == recipe.name for existing_recipe in self.recipe_list):
      self.recipe_list.append(recipe)
      for ingredient_obj in recipe.ingredient_dictionary.keys():  # Extract ingredient objects
        self.add_ingredient_to_list(ingredient_obj)
      self.save_data()
    else:
      print(f"Recipe '{recipe.name}' already exists in the list.")

  def add_ingredient_to_list(self, ingredient_obj):
    if not isinstance(ingredient_obj, Ingredient):
      return

    """Adds an ingredient object to the list if it doesn't already exist by name."""
    if not any(existing_ingredient.get_name() == ingredient_obj.get_name() for existing_ingredient in self.ingredient_list):
      self.ingredient_list.append(ingredient_obj)
      self.save_data()
    else:
      print(f"Ingredient '{ingredient_obj.get_name()}' already exists in the list.")

  def save_data(self):
    """Saves recipes and ingredients to a JSON file."""
    os.makedirs(os.path.dirname(self.DATA_FILE), exist_ok=True)  # Ensure /data directory exists

    data = {
      "recipes": [recipe.to_dict() for recipe in self.recipe_list],
      "ingredients": [ingredient.to_dict() for ingredient in self.ingredient_list]
    }

    with open(self.DATA_FILE, "w") as f:
      json.dump(data, f, indent=4)
    print("Data saved successfully.")

  def load_data(self):
    """Loads recipes and ingredients from a JSON file."""
    if os.path.exists(self.DATA_FILE):
      with open(self.DATA_FILE, "r") as f:
        data = json.load(f)

      self.recipe_list = [Recipe.from_dict(recipe) for recipe in data["recipes"]]
      self.ingredient_list = [Ingredient.from_dict(ingredient) for ingredient in data["ingredients"]]
      print("Data loaded successfully.")

  def get_all_recipes(self):
    """Returns the list of stored recipes."""
    return self.recipe_list

  def get_all_ingredients(self):
    """Returns the list of stored ingredients."""
    return self.ingredient_list
